File: examples/src/bin/status_check.rs
// Watch the EtherCAT link status for every device by driving the state checker.
// Run with: cargo xtask example status_check

use std::time::Duration;

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry};
use autd3_rs::{Client, ClientConfig, LinkStatus};
use autd3_rs_link_ethercrab::EtherCrabLinkOption;

const CHECK_INTERVAL: Duration = Duration::from_millis(100);

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("info")),
        )
        .init();

    let geometry = Geometry::new(vec![Autd3::default()]);

    let (client, mut checker) = Client::open_with_checker(
        &geometry,
        EtherCrabLinkOption::default(),
        ClientConfig::default(),
    )
    .await?;

    println!("watching link status — press Ctrl+C to stop");
    let mut last: Option<LinkStatus> = None;
    loop {
        let status = checker.check().await?;
        if last.as_ref() != Some(&status) {
            print_status(&status);
            last = Some(status);
        }
        tokio::select! {
            () = tokio::time::sleep(CHECK_INTERVAL) => {}
            _ = tokio::signal::ctrl_c() => break,
        }
    }

    client.close().await?;
    Ok(())
}

fn print_status(status: &LinkStatus) {
    for (i, state) in status.devices.iter().enumerate() {
        println!("device[{i}]: {state}");
    }
    println!(
        "all operational: {}, any lost: {}, recoveries: {}",
        status.all_op(),
        status.any_lost(),
        status.recoveries
    );
}

File: bindings/python/examples/status_check.py
"""
Watch the EtherCAT link status for every device.

Run with: cargo xtask py example status_check
"""

import asyncio
import signal

import autd3
import autd3_link_echocat as echocat

CHECK_INTERVAL = 0.1


async def main() -> None:
    geometry = autd3.geometry.Geometry([autd3.geometry.Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client, checker = await autd3.Client.open_with_checker(
        geometry,
        echocat.EchocatLinkOption(),
        autd3.ClientConfig(),
    )

    async with client:
        print("watching link status — press Ctrl+C to stop")
        stop = asyncio.Event()
        loop = asyncio.get_running_loop()
        for sig in (signal.SIGINT, signal.SIGTERM):
            loop.add_signal_handler(sig, stop.set)

        last = None
        while not stop.is_set():
            status = checker.check()
            key = (tuple(status.device_states), status.recoveries)
            if key != last:
                for i, state in enumerate(status.device_states):
                    print(f"device[{i}]: {state}")
                print(
                    f"all operational: {status.all_op}, "
                    f"any lost: {status.any_lost}, recoveries: {status.recoveries}"
                )
                last = key
            try:
                await asyncio.wait_for(stop.wait(), timeout=CHECK_INTERVAL)
            except asyncio.TimeoutError:
                pass


if __name__ == "__main__":
    asyncio.run(main())
